File: mobile/modules/core/android/src/main/java/com/mentra/core/ObservableStore.kt
package com.mentra.core

import org.json.JSONObject
import kotlinx.coroutines.*

/** Observable state management with debounced event emission */
class ObservableStore {
    private val values = mutableMapOf<String, Any>()
    private var onEmit: ((String, Map<String, Any>) -> Unit)? = null
    private val pendingChanges = mutableMapOf<String, MutableMap<String, Any>>()
    private var debounceJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    fun configure(onEmit: (String, Map<String, Any>) -> Unit) {
        this.onEmit = onEmit
    }

    private fun deepEquals(a: Any?, b: Any?): Boolean {
        return when {
            a === b -> true
            a == null || b == null -> false
            a is List<*> && b is List<*> -> a == b // List.equals does content comparison
            a is Map<*, *> && b is Map<*, *> -> a == b
            else -> a == b
        }
    }

    private fun toJson(value: Any): String {
        return JSONObject(mapOf("v" to value)).toString()
    }

    fun set(category: String, key: String, value: Any) {
        val fullKey = "$category.$key"
        val oldValue = values[fullKey]

        // Skip if unchanged (with deep equality for lists/maps)
        if (oldValue != null && toJson(oldValue) == toJson(value)) return

        values[fullKey] = value

        // Accumulate changes per category
        pendingChanges.getOrPut(category) { mutableMapOf() }[key] = value

        // Debounce emit (16ms ≈ 1 frame at 60fps)
        debounceJob?.cancel()
        debounceJob =
                scope.launch {
                    delay(16)
                    flushChanges()
                }
    }

    fun get(category: String, key: String): Any? = values["$category.$key"]

    fun getCategory(category: String): Map<String, Any> {
        val prefix = "$category."
        return values.filterKeys { it.startsWith(prefix) }.mapKeys { it.key.removePrefix(prefix) }
    }

    private fun flushChanges() {
        pendingChanges.forEach { (category, changes) ->
            if (changes.isNotEmpty()) {
                onEmit?.invoke(category, changes.toMap())
            }
        }
        pendingChanges.clear()
    }
}

File: mobile/modules/core/ios/Source/ObservableStore.swift
//
//  ObservableStore.swift
//  Core
//
//  Observable state management with debounced event emission
//

import Foundation

@MainActor
class ObservableStore {
    private var values: [String: Any] = [:]
    private var onEmit: ((String, [String: Any]) -> Void)?
    private var pendingChanges: [String: [String: Any]] = [:]
    private var debounceTask: Task<Void, Never>?

    func configure(onEmit: @escaping (String, [String: Any]) -> Void) {
        self.onEmit = onEmit
    }

    func set(_ category: String, _ key: String, _ value: Any) {
        let fullKey = "\(category).\(key)"
        let oldValue = values[fullKey]

        // Skip if unchanged (compare primitive types)
        if let old = oldValue, areEqual(old, value) {
            return
        }

        values[fullKey] = value

        // Accumulate changes per category
        if pendingChanges[category] == nil {
            pendingChanges[category] = [:]
        }
        pendingChanges[category]?[key] = value

        // Debounce emit
        debounceTask?.cancel()
        debounceTask = Task {
            try? await Task.sleep(nanoseconds: 16_000_000) // ~1 frame at 60fps
            guard !Task.isCancelled else { return }
            flushChanges()
        }
    }

    func get(_ category: String, _ key: String) -> Any? {
        values["\(category).\(key)"]
    }

    func getCategory(_ category: String) -> [String: Any] {
        var result: [String: Any] = [:]
        let prefix = "\(category)."
        for (key, value) in values where key.hasPrefix(prefix) {
            let shortKey = String(key.dropFirst(prefix.count))
            result[shortKey] = value
        }
        return result
    }

    private func flushChanges() {
        for (category, changes) in pendingChanges where !changes.isEmpty {
            onEmit?(category, changes)
        }
        pendingChanges.removeAll()
    }

    // Helper to compare values
    private func areEqual(_ lhs: Any, _ rhs: Any) -> Bool {
        // Handle primitive types
        if let l = lhs as? String, let r = rhs as? String { return l == r }
        if let l = lhs as? Int, let r = rhs as? Int { return l == r }
        if let l = lhs as? Bool, let r = rhs as? Bool { return l == r }
        if let l = lhs as? Double, let r = rhs as? Double { return l == r }
        return false
    }
}
